web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,439 @@
+import { AfterViewInit, Component, ElementRef, Input, ViewChild } from '@angular/core';
+import { HumanizeDurationLanguage, HumanizeDuration, HumanizeDurationOptions } from 'humanize-duration-ts';
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+
+/**
+ * Item holds all required data for one interval on the timeline.
+ */
+interface Item {
+  pushID: string;     // Push ID number
+  state: number;      // Final state of push
+  startTime: number;  // Start of push, in milliseconds
+  endTime: number;    // End of push, in milliseconds
+  row: number;        // Row number corresponds to y-position on timeine
+}
+
+@Component({
+  selector: 'app-timeline',
+  templateUrl: './timeline.component.html',
+  styleUrls: ['./timeline.component.scss']
+})
+
+export class TimelineComponent implements AfterViewInit {
+  /**
+   * Constants.
+   */
+  private static readonly LANG_SERVICE: HumanizeDurationLanguage = new HumanizeDurationLanguage();
+  private static readonly HUMANIZER: HumanizeDuration = new HumanizeDuration(TimelineComponent.LANG_SERVICE);
+  private static readonly COLOR_LIGHT_GRAY: string = '#d3d3d3';
+  private static readonly MIN_INTERVAL_HEIGHT: number = 25;
+  private static readonly MSEC_PER_MIN: number = 60 * (10 ** 3);
+  private static readonly NSEC_PER_MSEC: number = 10 ** 6;
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    1: '#eee',
+    3: '#2196f3',
+    4: '#d50000',
+    5: '#34a853',
+    6: '#d50000',
+    7: '#2196f3',
+    8: '#2196f3',
+    9: '#d50000',
+    10: '#2196f3',
+    11: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    14: '#eee',
+    15: '#2196f3',
+    16: '#d50000',
+    17: '#eee',
+    18: '#34a853',
+    19: '#eee'
+  };
+
+  /**
+   * Private data variables.
+   *
+   * Note that we use the non-null assertion operator ('!') to indicate to
+   * the compiler that variables will always be declared and thus to not
+   * issue errors about possibilities of them being `null` or `undefined`
+   * since pushInfos is a passed-in value.
```

Comment:
I think we can drop this because the `@ViewChild` and `@Input` are using `!` now.